products/bridge/smart-contracts/contracts/periphery/TokenManagerV2/TokenManagerUpgradeableV2.sol

```diff
@@ -0,0 +1,192 @@
+    bytes32 private constant Token_Manager_Storage_Location =
+        0x4a6c2e6a7e6518c249bdcd1d934ea16ea5325bbae105af814eb678f5f49f3400;
+
+    function _getTokenManagerStorage()
+        private
+        pure
+        returns (TokenManagerStorage storage $)
+    {
+        assembly {
+            $.slot := Token_Manager_Storage_Location
+        }
+    }
+
+    function getGateway() public view returns (address) {
+        TokenManagerStorage storage $ = _getTokenManagerStorage();
+        return $.gateway;
+    }
+
+    function getRemoteTokens(
+        address token,
+        uint remoteChainId
+    ) public view returns (RemoteToken memory) {
+        TokenManagerStorage storage $ = _getTokenManagerStorage();
+        return $.remoteTokens[token][remoteChainId];
+    }
+
+    modifier onlyGateway() {
+        if (msg.sender != address(getGateway())) {
+            revert NotGateway();
+        }
+        _;
+    }
+
+    function __TokenManager_init(address _gateway) internal onlyInitializing {
+        __Ownable_init(msg.sender);
+        _setGateway(_gateway);
+    }
+
+    function _authorizeUpgrade(address) internal virtual override onlyOwner {}
+
+    function _setGateway(address _gateway) internal {
+        TokenManagerStorage storage $ = _getTokenManagerStorage();
+        $.gateway = _gateway;
+    }
+
+    function setGateway(address _gateway) external onlyOwner {
+        _setGateway(_gateway);
+    }
+
+    function _removeToken(address localToken, uint remoteChainId) internal {
+        TokenManagerStorage storage $ = _getTokenManagerStorage();
+        delete $.remoteTokens[localToken][remoteChainId];
+        emit TokenRemoved(localToken, remoteChainId);
+    }
+
+    function _registerToken(
+        address localToken,
+        RemoteToken memory remoteToken
+    ) internal {
+        TokenManagerStorage storage $ = _getTokenManagerStorage();
+        $.remoteTokens[localToken][remoteToken.chainId] = remoteToken;
+        emit TokenRegistered(
+            localToken,
+            remoteToken.token,
+            remoteToken.tokenManager,
+            remoteToken.chainId
+        );
+    }
+
+    // Token Overrides
+    function registerToken(
+        address token,
+        RemoteToken memory remoteToken
+    ) external virtual onlyOwner {
+        _registerToken(token, remoteToken);
+    }
+
+    function setFees(uint newFees) external override onlyOwner {
+        _setFees(newFees);
+    }
+
+    // TO OVERRIDE – Incoming
+    function _handleTransfer(
+        address token,
+        address from,
+        uint amount
+    ) internal virtual;
+
+    // TO OVERRIDE – Outgoing
+    function _handleAccept(
+        address token,
+        address recipient,
+        uint amount
+    ) internal virtual;
+
+    function transfer(
+        address token,
+        uint remoteChainId,
+        address remoteRecipient,
+        uint amount
+    ) external payable virtual checkFees {
```

Comment:
`checkFees` is the critical line to asserting fees and ensure the contract is payable.